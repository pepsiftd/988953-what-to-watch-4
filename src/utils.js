const getUniqueItems = (list) => {
  return list.filter((it, i, array) => {
    return array.indexOf(it) === i;
  });
};

const extend = (a, b) => {
  return Object.assign({}, a, b);
};

const getMoviesByGenre = (movies, genre) => {
  return genre === `All genres` ? movies : movies.filter((movie) => movie.genre === genre);
};

const getGenresFromMovies = (movies) => {
  return getUniqueItems(movies.map((movie) => movie.genre));
};

export {extend, getMoviesByGenre, getUniqueItems, getGenresFromMovies};
