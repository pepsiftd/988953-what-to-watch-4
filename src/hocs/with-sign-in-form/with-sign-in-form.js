import React, {PureComponent, createRef} from 'react';
import PropTypes from 'prop-types';

import {checkEmail} from '@/utils';

const withSignInForm = (Component) => {
  class WithSignInForm extends PureComponent {
    constructor(props) {
      super(props);

      this._emailInputRef = createRef();
      this._passwordInputRef = createRef();

      this.state = {
        isValidEmail: true,
      };

      this.onSubmit = this.onSubmit.bind(this);
      this.onEmailInput = this.onEmailInput.bind(this);
    }

    onEmailInput() {
      if (checkEmail(this._emailInputRef.current.value)) {
        this.setState({
          isValidEmail: true
        });
      } else {
        this.setState({
          isValidEmail: false
        });
      }
    }

    onSubmit(evt) {
      evt.preventDefault();

      if (this.state.isValidEmail && this._passwordInputRef.current.value.length > 0) {
        this.props.onSignIn({
          email: this._emailInputRef.current.value,
          password: this._passwordInputRef.current.value,
        });
      }
    }

    render() {
      return (
        <Component
          {...this.props}
          renderEmailInput={() => {
            return (
              <input
                ref={this._emailInputRef}
                className="sign-in__input"
                type="email"
                placeholder="Email address"
                name="user-email"
                id="user-email"
                onInput={this.onEmailInput}
              />
            );
          }}

          renderPasswordInput={() => {
            return (
              <input
                ref={this._passwordInputRef}
                className="sign-in__input"
                type="password"
                placeholder="Password"
                name="user-password"
                id="user-password"
              />
            );
          }}

          isValidEmail={this.state.isValidEmail}
          onSubmit={this.onSubmit}
        />
      );
    }
  }

  WithSignInForm.propTypes = {
    onSignIn: PropTypes.func.isRequired,
  };

  return WithSignInForm;
};

export {withSignInForm};
