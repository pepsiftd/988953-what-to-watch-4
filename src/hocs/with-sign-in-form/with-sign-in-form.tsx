import React, {PureComponent, createRef} from 'react';
import {Subtract} from 'utility-types';

import {checkEmail} from '@/utils';

interface State {
  isValidEmail: boolean;
}

interface InjectingProps {
  renderEmailInput: () => React.ReactNode;
  renderPasswordInput: () => React.ReactNode;
  isValidEmail: boolean;
  onSubmit: (evt: React.FormEvent) => void;
}

const withSignInForm: (Component: React.ComponentClass) => React.ReactNode = (Component) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Subtract<P, InjectingProps>;

  class WithSignInForm extends PureComponent<T, State> {
    private emailInputRef: React.RefObject<HTMLInputElement>;
    private passwordInputRef: React.RefObject<HTMLInputElement>;

    constructor(props) {
      super(props);

      this.emailInputRef = createRef();
      this.passwordInputRef = createRef();

      this.state = {
        isValidEmail: true,
      };

      this.onSubmit = this.onSubmit.bind(this);
      this.onEmailInput = this.onEmailInput.bind(this);
    }

    onEmailInput() {
      if (checkEmail(this.emailInputRef.current.value)) {
        this.setState({
          isValidEmail: true
        });
      } else {
        this.setState({
          isValidEmail: false
        });
      }
    }

    onSubmit(evt) {
      evt.preventDefault();

      if (this.state.isValidEmail && this.passwordInputRef.current.value.length > 0) {
        this.props.onSignIn({
          email: this.emailInputRef.current.value,
          password: this.passwordInputRef.current.value,
        });
      }
    }

    render() {
      return (
        <Component
          {...this.props}
          renderEmailInput={() => {
            return (
              <input
                ref={this.emailInputRef}
                className="sign-in__input"
                type="email"
                placeholder="Email address"
                name="user-email"
                id="user-email"
                onInput={this.onEmailInput}
              />
            );
          }}

          renderPasswordInput={() => {
            return (
              <input
                ref={this.passwordInputRef}
                className="sign-in__input"
                type="password"
                placeholder="Password"
                name="user-password"
                id="user-password"
              />
            );
          }}

          isValidEmail={this.state.isValidEmail}
          onSubmit={this.onSubmit}
        />
      );
    }
  }

  return WithSignInForm;
};

export {withSignInForm};
