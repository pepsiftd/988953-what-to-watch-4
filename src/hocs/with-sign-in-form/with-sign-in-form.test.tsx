import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';
import {noop} from '@/utils';

import {withSignInForm} from './with-sign-in-form';

const MockComponent = ({renderEmailInput, renderPasswordInput, isValidEmail, onSubmit}) => {
  return (
    <form
      onSubmit={onSubmit}
    >
      {isValidEmail && <p>Email is valid</p>}
      {renderEmailInput()}
      {renderPasswordInput()}
    </form>
  );
};

MockComponent.propTypes = {
  renderPasswordInput: PropTypes.func.isRequired,
  renderEmailInput: PropTypes.func.isRequired,
  isValidEmail: PropTypes.bool.isRequired,
  onSubmit: PropTypes.func.isRequired,
};


it(`withSignInForm renders inputs correctly`, () => {
  const MockComponentWrapped = withSignInForm(MockComponent);

  const tree = renderer.create(
      <MockComponentWrapped
        onSignIn={noop}
      />
  );

  expect(tree).toMatchSnapshot();
});
