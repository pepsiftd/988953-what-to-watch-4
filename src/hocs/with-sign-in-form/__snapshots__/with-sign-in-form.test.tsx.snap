// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSignInForm renders inputs correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <p>
    Email is valid
  </p>
  <input
    className="sign-in__input"
    id="user-email"
    name="user-email"
    onInput={[Function]}
    placeholder="Email address"
    type="email"
  />
  <input
    className="sign-in__input"
    id="user-password"
    name="user-password"
    placeholder="Password"
    type="password"
  />
</form>
`;
